sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -787,6 +788,28 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+ * Simplifies semantically-equal non-nullable expressions:
+ * 1) Replace '=', '<=', and '>=' with 'true' literal
+ * 2) Replace '<' and '>' with 'false' literal
+ */
+object NonNullableBinaryComparisonSimplification extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      // True with equality
+      case a EqualTo b if !a.nullable && !b.nullable && a.semanticEquals(b) => TrueLiteral
+      case a EqualNullSafe b if !a.nullable && !b.nullable && a.semanticEquals(b) => TrueLiteral
```

Comment:
The javadoc of `semanticEquals` says: `returns true when two expressions will always compute the same result`. So I think they should be both null or both not null if they are `semanticEquals`.